v3/impl.h

```diff
@@ -210,28 +213,60 @@ struct wide_integer<Bits, Signed>::_impl {
     }
 
     constexpr static void wide_integer_from_bultin(wide_integer<Bits, Signed>& self, double rhs) noexcept {
-        if ((rhs > 0 && rhs < std::numeric_limits<uint64_t>::max()) ||
-            (rhs < 0 && rhs > std::numeric_limits<int64_t>::min())) {
+        constexpr int64_t max_int = std::numeric_limits<int64_t>::max();
+        constexpr int64_t min_int = std::numeric_limits<int64_t>::min();
+
+        constexpr long double max_int_long_double = static_cast<long double>(max_int);
+
+        if ((rhs > 0 && rhs < max_int) ||
+            (rhs < 0 && rhs > min_int))
+        {
             self = to_Integral(rhs);
             return;
         }
 
-        long double r = rhs;
-        if (r < 0) {
-            r = -r;
+        /// There are values in int64 that have more than 53 significant bits (in terms of double
+        /// representation). Such values, being promoted to double, are rounded up or down. If they are rounded up,
+        /// the result may not fit in 64 bits.
+        /// The example of such a number is 9.22337e+18.
+        /// As to_Integral does a static_cast to int64_t, it may result in UB.
+        /// The necessary check here is that long double has enough significant (mantissa) bits to store the
+        /// int64_t max value precisely.
+        static_assert(LDBL_MANT_DIG >= 64,
+            "On your system long double has less than 64 precision bits,"
+            "which may result in UB when initializing double from int64_t");
+
+        /// Always >= 0
+        const long double rhs_long_double = (static_cast<long double>(rhs) < 0)
+            ? -static_cast<long double>(rhs)
+            : rhs;
+
+        const long double rhs_max_int_count = rhs_long_double / max_int;
+
+        // We can't just get the number of iterations like rhs_max_int_count / max_int as it may not fit it int64_t.
+        long double rhs_max_int_count_acc = rhs_max_int_count;
+
+        self = 0;
+
+        while (rhs_max_int_count_acc > max_int_long_double)
+        {
+            self += max_int;
+            rhs_max_int_count_acc -= max_int_long_double;
```

Comment:
This loop could be extremely slow.

Maybe `std::ilogb()`+`std::scalbn()` could help to fill `self` with some numerable iterations?

Reply:
This still is not an optimal solution, I prefer to think of it as a more-or-less working one,

Reply:
Turns out this while inappropriately slowed down everything.